### ISSUE TITLE
How to catch ruby Exceptions in a c extension

### ISSUE THREAD

**Issue by Asmod4n** [mruby/mruby#2837]
What is the proper way to use MRB_TRY / CATCH in a c extension?
Currently have to store temporary state in a mrb_gv which would normally be on the stack.

**Comment by beoran** [comment]
Those macros are for mruby's internal use. To catch a mruby exception simply check the mrb->exc member. If it's not NULL it contains the last exception raised. Then you can use mrb_funcall(mrb, mrb_obj_value(mrb->exc), ""backtrace", 0) to get the backtrace of that exception.

**Comment by Asmod4n** [comment]
@beoran 
When the block which gets called here https://github.com/Asmod4n/mruby-simplemsgpack/blob/master/src/mrb_msgpack.c#L183 raises a exception i never get the chance to check for mrb->exc, the loop ends and the function gets jumped out, or am i missing something here?

**Comment by beoran** [comment]
Oh, you want to catch exceptions raised from an mrb_ yield? AFAIK that's not supported yet, but I could be mistaken. Maybe someone else knows? @matz , could you enlighten us please?

**Comment by Asmod4n** [comment]
What i basically need is a ensure "block", no matter what happens, i allways have to call ```msgpack_unpacked_destroy``` at the end of the c func or memory leaks happen.

**Comment by matz** [comment]
Maybe we should provide `mrb_protect()` as CRuby does.

**Comment by beoran** [comment]
Yes, please. :-)

**Comment by Asmod4n** [comment]
yay :)

**Comment by thomthom** [comment]
> Those macros are for mruby's internal use. To catch a mruby exception simply check the mrb->exc member. If it's not NULL it contains the last exception raised. Then you can use mrb_funcall(mrb, mrb_obj_value(mrb->exc), ""backtrace", 0) to get the backtrace of that exception.

Is there anything that needs to be reset after an exception is thrown? It appear that if I get and exception during `mrb_load_nstring_cxt` consecutive calls will fail.

**Comment by matz** [comment]
`mrb->exc = NULL;` should work. Probably we'd better prepare API functions for the purpose.

**Comment by thomthom** [comment]
That did indeed work. I tried that yesterday after digging into mruby's own source code and saw that being done. Though i wasn't sure if anything else should be set.

I also saw that `mrb_protect` exist. Though I got linking errors when trying to use it. I guess that's something I need to explicitly include when building Ruby? I also wasn't entirely use how to use it.

I am familiar with Ruby's C API however, I was trying to refer to how protect is used there. But even there I've never been too sure on how to handle the protect wrappers when I have multiple arguments. Any docs or examples that show best practice on this?

So far I've been referring to this guide for the C API (with vanilla Ruby):
http://silverhammermba.github.io/emberb/c/

That give a rather nice breakdown of common how to's. Anything similar to mruby?
